url.go

```diff
@@ -1,35 +1,240 @@
 package s3client
 
 import (
+	"errors"
+	"fmt"
 	"net/url"
 	"strings"
 )
 
-// NewURL create a new instance of URL from a virtual-hosted-style URL
-// e.g. 'https://bucktName/ObjectKey'
-func NewURL(rawURL string) (*URL, error) {
+// URLStyle is the type to define the URL style iota enumeration corresponding an S3 url (path, virtualHosted, etc)
+type URLStyle int
 
-	url, err := url.Parse(rawURL)
+// Possible S3 URL styles
+const (
+	StylePath = iota
+	StyleGlobalPath
+	StyleVirtualHosted
+	StyleGlobalVirtualHosted
+	StyleAliasVirtualHosted
+)
+
+var urlStyles = []string{"Path", "GlobalPath", "VirtualHosted", "GlobalVirtualHosted", "AliasVirtualHosted"}
+
+// Values of the format styles
+func (style URLStyle) String() string {
+	return urlStyles[style]
+}
+
+// S3Url represents an S3 URL with bucketName, key and region (optional). This struct is
+// intended to be used for S3 URL string manipulation/translation in its possible format styles.
+type S3Url struct {
+	Scheme     string
+	Region     string
+	BucketName string
+	Key        string
+}
+
+// String returns the S3 URL string in the requested format style.
+// Possible formats are defined in https://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html
```

Comment:
I would add these comments (example locations) against the constants at the top of page

Reply:
Also some of the names do not match up PathStype and StylePath etc. - please double check all of them thanks